### PR TITLE
[reliable payments] Move payment error deobfuscation to router

### PR DESCRIPTION
**Pull request by halseth**
This PR consists of two important pieces of #2761: 

### Make `router`<->`switch` interaction async
We define `PaymentResult` and the  method `GetPaymentResult`
exposed by the `switch`.

This lets us distinguish an critical error from a actual payment result
(success or failure). This is important since we know that we can only
attempt another payment when a final result from the previous payment
attempt is received.

In order to achieve this we let the `router` handling generating `paymentID`s,
since they are used to later query the `switch` for the final result.

### Move payment error deobfuscation
In this commit we move handing the deobfuscator from the router to the
switch from when the payment is initiated, to when the result is
queried.

We do this because only the router can recreate the deobfuscator after a
restart, and we are preparing for being able to handle results across
restarts.

Since the deobfuscator cannot be nil anymore, we can also get rid of
that special case.

Split off from #2761

### CODE REVIEW

File: routing/router.go [hunk 1]
```diff
@@ -180,7 +179,15 @@ type Config struct {
...
 		htlcAdd *lnwire.UpdateAddHTLC,
-		circuit *sphinx.Circuit) ([sha256.Size]byte, error)
+		circuit *sphinx.Circuit) error
+
+	// GetPaymentResult returns the the result of the payment attempt with
+	// the given paymentID. The method returns a channel where the payment
+	// result will be sent when available, or an error is encountered. If
+	// the paymentID is unknown, htlcswitch.ErrPaymentIDNotFound will be
+	// returned.
+	GetPaymentResult func(paymentID uint64) (
```
Comment (Roasbeef) [inline routing/router.go]: Not a blocker, but at this point it may make sense to extract this method and the one above it into a new interface.

File: channeldb/graph_test.go [hunk 2]
```diff
@@ -3005,3 +3009,54 @@ func compareEdgePolicies(a, b *ChannelEdgePolicy) error {
 	}
 	return nil
 }
+
+// TestLightningNodeSigVerifcation checks that we can use the LightningNode's
+// pubkey to verify signatures.
```
Comment (Roasbeef) [inline channeldb/graph_test.go]: This commit can be removed now.

File: htlcswitch/link_test.go [hunk 3]
```diff
@@ -1110,7 +1110,6 @@ func TestChannelLinkMultiHopUnknownPaymentHash(t *testing.T) {
 	// Send payment and expose err channel.
 	err = n.aliceServer.htlcSwitch.SendHTLC(
 		n.firstBobChannelLink.ShortChanID(), pid, htlc,
-		newMockDeobfuscator(),
```
Comment (Roasbeef) [inline htlcswitch/link_test.go]: Is this movement really needed in order to complete the reliable payments feature? The onion decryption/encryption portion is a _link level_ detail, meaning that the router doesn't need to be aware of it at all. The current placement is by design, in that the router crafts the route, but all error encryption/decryption is handled by the switch. As a result, the router gets back a nice concrete error, and doesn't need to know about the details of error encryption/decryption. 

IMO, it's a code smell that the router is now _creating_ then _returning_ an `htlcswitch.ForwardingError` back to itself. It's also the case that we already store the items required to reconstruct the circuit (for encrypting and decrypting errors) in the switch's circuit map. Additionally, there's some particular context in the switch w.r.t things like if this is a locally sourced error, a resolution, and so on that are required in order to properly handle all classes of errors. Overall, I don't see why this change is warranted.

Comment (halseth) [reply htlcswitch/link_test.go]: As pointed out in https://github.com/lightningnetwork/lnd/pull/2762/#discussion_r281499203, the current placement is awkward, as _both_ the `router` (encryption) and `switch` (decryption) is aware of the onion format now. By moving it all into either one, it should be less shared responsibility.

I moved it all into the `router` since it has everything that is needed to decrypt the error (route + session key) by recreating the `ErrorDecryptor` after a restart. We could possibly move it all into the `switch`, but that would require us to store this information across restarts, as it is currently not (the curcuit map only stores onion blobs from what I can tell). If that's what we want then a natural place to put it would be the circuit map, as part of the `paymentCircuit`?

And finally, the reason this is needed for reliable payments is to handle received errors across restarts. The `switch` will already handle persisting received results (see #2762), but it cannot deobfuscate the errors. This has to be done by the router, either by getting the obfuscated data from the switch, or by giving it the `deobfuscator` to the switch after a restart (but this is a bit tricky timingwise).

File: channeldb/graph_test.go (hunk 2 above)
Comment (halseth) [reply channeldb/graph_test.go]: Done

File: routing/router.go (hunk 1 above)
Comment (halseth) [reply routing/router.go]: Made interface

File: htlcswitch/link_test.go (hunk 3 above)
Comment (halseth) [reply htlcswitch/link_test.go]: > or by giving it the deobfuscator to the switch after a restart (but this is a bit tricky timingwise).
I made a rough attempt doing this, and the diff is not that bad:  https://github.com/lightningnetwork/lnd/commit/9e2c428f34c6de9486fceb4d09a376671b4e578c

lmk what is preferred.

Comment (halseth) [reply htlcswitch/link_test.go]: Went with the smallest diff case.

**Comment by halseth** [comment]
PTAL @Roasbeef

File: routing/router.go [hunk 4]
```diff
@@ -1694,9 +1704,67 @@ func (r *ChannelRouter) sendToSwitch(route *route.Route, paymentHash [32]byte) (
...
+	err = r.cfg.Payer.SendHTLC(
+		firstHop, paymentID, htlcAdd,
+	)
+	if err != nil {
+		log.Errorf("Failed sending attempt %d for payment %x to "+
+			"switch: %v", paymentID, paymentHash, err)
+
+		// We must inspect the error to know whether it was critical or
+		// not, to decide whether we should continue trying.
+		finalOutcome := r.processSendError(
```
Comment (joostjager) [inline routing/router.go]: This is not the type of error that is suitable for `processSendError`. It will never be a wire error at this stage and `finalOutcome` is always `true`?

File: htlcswitch/switch.go [hunk 5]
```diff
@@ -353,35 +340,80 @@ func (s *Switch) ProcessContractResolution(msg contractcourt.ResolutionMsg) erro
...
+			return
+		}
+
+		// Extract the result and pass it to the result channel.
+		result, err := s.extractResult(
+			deobfuscator, n, payment.paymentHash,
+		)
+		if err != nil {
+			log.Errorf("Unable to extract result: %v", err)
+			return
```
Comment (joostjager) [inline htlcswitch/switch.go]: This will leave the `router` blocked?

File: htlcswitch/switch.go [hunk 6]
```diff
@@ -889,64 +903,80 @@ func (s *Switch) handleLocalResponse(pkt *htlcPacket) {
 	// has been restarted since sending the payment.
 	payment := s.findPayment(pkt.incomingHTLCID)
```
Comment (joostjager) [inline htlcswitch/switch.go]: If `payment` is nil, we can return already here.

File: htlcswitch/switch.go [hunk 7]
```diff
@@ -990,29 +1019,16 @@ func (s *Switch) parseFailedPayment(payment *pendingPayment, pkt *htlcPacket,
...
-	// channel failure to the router.
-	case payment == nil:
-		userErr := fmt.Sprintf("error decryptor for payment " +
-			"could not be located, likely due to restart")
-		failure = &ForwardingError{
-			ErrorSource:    s.cfg.SelfKey,
-			ExtraMsg:       userErr,
-			FailureMessage: lnwire.NewTemporaryChannelFailure(nil),
-		}
-
```
Comment (joostjager) [inline htlcswitch/switch.go]: :+1:

File: htlcswitch/switch.go [hunk 8]
```diff
@@ -956,14 +986,13 @@ func (s *Switch) parseFailedPayment(payment *pendingPayment, pkt *htlcPacket,
...
 		var userErr string
 		r := bytes.NewReader(htlc.Reason)
 		failureMsg, err := lnwire.DecodeFailure(r, 0)
 		if err != nil {
-			userErr = fmt.Sprintf("unable to decode onion failure, "+
-				"htlc with hash(%x): %v",
-				pkt.circuit.PaymentHash[:], err)
+			userErr = fmt.Sprintf("unable to decode onion "+
+				"failure: %v", err)
 			log.Error(userErr)
```
Comment (joostjager) [inline htlcswitch/switch.go]: Is the log line useful without any identifying info? Same for other statement in this function.

File: routing/mock_test.go [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+		return nil, htlcswitch.ErrPaymentIDNotFound
+	}
+	c <- res
+
+	return c, nil
+
+}
+
+func (m *mockPaymentAttemptDispatcher) setPaymentResult(f func(firstHop lnwire.ShortChannelID) ([32]byte, error)) {
+
```
Comment (joostjager) [inline routing/mock_test.go]: newline

File: routing/router.go [hunk 10]
```diff
@@ -1694,9 +1704,67 @@ func (r *ChannelRouter) sendToSwitch(route *route.Route, paymentHash [32]byte) (
 	firstHop := lnwire.NewShortChanIDFromInt(
 		route.Hops[0].ChannelID,
 	)
-	return r.cfg.SendToSwitch(
-		firstHop, htlcAdd, circuit,
+
+	// Using the created circuit, initialize the error
+	// decrypter so we can parse+decode any failures
+	// incurred by this payment within the switch.
+	errorDecryptor := &htlcswitch.SphinxErrorDecrypter{
```
Comment (joostjager) [inline routing/router.go]: Can be moved down to where it is used.

File: htlcswitch/switch.go [hunk 11]
```diff
@@ -889,64 +903,80 @@ func (s *Switch) handleLocalResponse(pkt *htlcPacket) {
...
+		unencrypted:  unencrypted,
+		isResolution: pkt.isResolution,
+	}
 
-	switch htlc := pkt.htlc.(type) {
+	// Deliver the payment error and preimage to the application, if it is
+	// waiting for a response.
+	if payment != nil {
+		payment.resultChan <- n
+		s.removePendingPayment(pkt.incomingHTLCID)
```
Comment (cfromknecht) [inline htlcswitch/switch.go]: seems like there's a possible race where:
1. `SendHTLC` is called, and the pending payment is added to the map
2. The HTLC is settled/failed (most likely, failed locally) and the pending payment is removed from the map
3. `GetPaymentResult` is called but the pending payment is no longer found

Perhaps we can just opt to not remove pending payments from the map? Would think the flow would be more like:
1. If pending payment exists, wait for ntfn to be delivered via channel
2. If no pending payment exists, retrieve from database or cached in-memory response

Comment (Roasbeef) [reply htlcswitch/switch.go]: Yeah this doesn't yet resolve races like this or gaps in persistence. Those will be taken care of in  #2762 (the pending payments map goes away all together).

File: htlcswitch/payment_result.go [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+	// msg is the received result. This should be of type UpdateFulfillHTLC
+	// or UpdateFailHTLC.
+	msg lnwire.Message
+
+	// unencrypted indicates whether the failure encoded in the message is
+	// unencrypted, and hence doesn't need to be decrypted.
+	unencrypted bool
+
+	// isResolution indicates whether this is a resolution message, in
+	// which the failure reason might not be included.
```
[the remark this answers is not in the document]
Comment (Roasbeef) [reply htlcswitch/payment_result.go]: IIRC we return "permanent chan failure" if we need to sweep it on chain, then cancel back along the incoming link.

File: htlcswitch/switch.go (hunk 8 above)
Comment (Roasbeef) [reply htlcswitch/switch.go]: Agree we should still log the payment hash at the very least.

File: routing/router.go (hunk 4 above)
Comment (Roasbeef) [reply routing/router.go]: Yeah I think this will be the case, as if we get an error here, then it never left the switch. Will need to double check before we modify this at all. But in the end, having this statement here may actually be harmless.

File: htlcswitch/switch.go (hunk 5 above)
Comment (Roasbeef) [reply htlcswitch/switch.go]: Yeah we need to return an error result in this case. Also above when we exit due to a quit channel.

File: htlcswitch/switch.go [hunk 13]
```diff
@@ -353,35 +340,80 @@ func (s *Switch) ProcessContractResolution(msg contractcourt.ResolutionMsg) erro
 	return nil
 }
 
+// GetPaymentResult returns the the result of the payment attempt with the
+// given paymentID. The method returns a channel where the payment result will
+// be sent when available, or an error is encountered. If the paymentID is
+// unknown, ErrPaymentIDNotFound will be returned.
+func (s *Switch) GetPaymentResult(paymentID uint64,
```
[the remark this answers is not in the document]
Comment (Roasbeef) [reply htlcswitch/switch.go]: More than once as in multiple subscribers?

File: htlcswitch/switch.go (hunk 11 above)
Comment (cfromknecht) [reply htlcswitch/switch.go]: gotcha, okay looking through #2762 this race does seem to be eliminated

File: routing/mock_test.go [hunk 14]
```diff
@@ -0,0 +1,30 @@
+package routing
```
[the remark this answers is not in the document]
Comment (halseth) [reply routing/mock_test.go]: Fixed.

File: routing/mock_test.go [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+	_ htlcswitch.ErrorDecrypter) ([sha256.Size]byte, error) {
+
+	if m.onPayment != nil {
+		return m.onPayment(firstHop)
+	}
+
+	return [sha256.Size]byte{}, nil
+}
+
+func (m *mockPaymentAttemptDispatcher) setPaymentResult(f func(firstHop lnwire.ShortChannelID) ([32]byte, error)) {
```
[the remark this answers is not in the document]
Comment (halseth) [reply routing/mock_test.go]: Fixed: https://github.com/lightningnetwork/lnd/pull/3063/commits/6e28fb8d374e818b8d8cb7b305ce5ec54dbf2190

File: routing/mock_test.go (hunk 9 above)
Comment (halseth) [reply routing/mock_test.go]: Fixed: https://github.com/lightningnetwork/lnd/pull/3063/commits/6e28fb8d374e818b8d8cb7b305ce5ec54dbf2190

File: routing/router.go [hunk 16]
```diff
@@ -1694,8 +1703,16 @@ func (r *ChannelRouter) sendToSwitch(route *route.Route, paymentHash [32]byte) (
 	firstHop := lnwire.NewShortChanIDFromInt(
 		route.Hops[0].ChannelID,
 	)
-	return r.cfg.SendToSwitch(
-		firstHop, htlcAdd, circuit,
+
+	// Using the created circuit, initialize the error
```
[the remark this answers is not in the document]
Comment (halseth) [reply routing/router.go]: Fixed: https://github.com/lightningnetwork/lnd/pull/3063/commits/9df97c56ddd378448def5065002123e91df2909d

File: routing/router.go (hunk 10 above)
Comment (halseth) [reply routing/router.go]: Not sure what you mean?

File: htlcswitch/switch.go (hunk 8 above)
Comment (halseth) [reply htlcswitch/switch.go]: In upcoming PRs we won't always have the payment hash available at this point (we would need to persist it). Added payment ID to the logs: https://github.com/lightningnetwork/lnd/pull/3063/commits/0062686efade9ac80de973caeb0b6b1f9181360a

File: htlcswitch/switch.go (hunk 6 above)
Comment (halseth) [reply htlcswitch/switch.go]: Parse failed payment will actually log. Also this will change in upcoming PR.

File: routing/router.go (hunk 4 above)
Comment (halseth) [reply routing/router.go]: We will return a `ForwardingError` for some local cases, like if no Link has capacity.

File: htlcswitch/switch.go (hunk 11 above)
Comment (halseth) [reply htlcswitch/switch.go]: It's a good point, it does introduce a race. This is fixed in #2762, but it is a bummer that this PR introduces the race. Temp fixed it by not removing the pending payment at all: https://github.com/lightningnetwork/lnd/pull/3063/commits/1f9b4fb260979920cb62eda3e9940384becdc1bd

File: htlcswitch/switch.go [hunk 17]
```diff
@@ -889,64 +903,80 @@ func (s *Switch) handleLocalResponse(pkt *htlcPacket) {
...
+	switch htlc := n.msg.(type) {
 
 	// We've received a settle update which means we can finalize the user
 	// payment and return successful response.
 	case *lnwire.UpdateFulfillHTLC:
 		// Persistently mark that a payment to this payment hash
 		// succeeded. This will prevent us from ever making another
 		// payment to this hash.
-		err := s.control.Success(pkt.circuit.PaymentHash)
+		err := s.control.Success(paymentHash)
```
[the remark this answers is not in the document]
Comment (halseth) [reply htlcswitch/switch.go]: Yeah, when the control tower is removed, it will be more aptly named.

File: htlcswitch/switch.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (halseth) [reply htlcswitch/switch.go]: It will support multiple callers, but I guess it won't happen.

File: htlcswitch/payment_result.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (halseth) [reply htlcswitch/payment_result.go]: I see at least one place where the failure reason is non-nil: https://github.com/lightningnetwork/lnd/blob/a8fa4094ba61e26bd9855e9743883e3a234d383a/htlcswitch/switch.go#L1184

cc @cfromknecht

File: htlcswitch/switch.go [hunk 18]
```diff
@@ -353,35 +340,80 @@ func (s *Switch) ProcessContractResolution(msg contractcourt.ResolutionMsg) erro
...
+		return nil, ErrPaymentIDNotFound
+	}
+
+	resultChan := make(chan *PaymentResult, 1)
+
+	// Since the payment was known, we can start a goroutine that can
+	// extract the result when it is available, and pass it on to the
+	// caller.
+	s.wg.Add(1)
+	go func() {
```
[the remark this answers is not in the document]
Comment (halseth) [reply htlcswitch/switch.go]: `GetPaymentID` needs to return for the router to know whether the payment ID existed (also to not deadlock the interaction router<->switch on shutdown).

File: htlcswitch/switch.go (hunk 5 above)
Comment (halseth) [reply htlcswitch/switch.go]: Hm, yeah this is interesting. When calling `GetPaymentResult` we need to distinguish between 3 types of errors:

1. An immediate error whether the subscription was successful (currently only error is in case of unknown payment ID).
2. Critical error that happened when the result was received (this case). This is not a `ForwardingError` (though we could wrap it in one?).
3. A forwarding error.

I see two ways of designing this:

* Make `PaymentResult` also hold critical errors:
```go
type PaymentResult struct {
	// Preimage is set by the switch in case a sent HTLC was settled.
	Preimage [32]byte

 	// Error is non-nil in case a HTLC send failed, and the HTLC is now
	// irrevocably cancelled.
	Error error
}
```
The reason this could only be `ForwardingErrors` is that in an earlier iteration this whole struct was serialized.
* Wrap the critical error in a `ForwardingError`.

File: htlcswitch/switch.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (joostjager) [reply htlcswitch/switch.go]: Yes, calling with the same pid twice.

File: htlcswitch/payment_result.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (joostjager) [reply htlcswitch/payment_result.go]: In that line, it isn't a locally sourced payment. So that reason can not end up as a payment result in the switch with `isResolution` true?

File: htlcswitch/switch.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (joostjager) [reply htlcswitch/switch.go]: But `GetPaymentResult` without the goroutine would still return early with `ErrPaymentIDNotFound` if it doesn't exist?

Does a dead lock happen if both router and switch wait for closure of the same quit chan?

File: htlcswitch/switch.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (joostjager) [reply htlcswitch/switch.go]: How would it support multiple callers if there is only one `resultChan`? Only the first callers receives the result and the next will wait forever?

File: htlcswitch/switch.go (hunk 5 above)
Comment (halseth) [reply htlcswitch/switch.go]: I realized that we cannot return any result in case of quit, so maybe we can just close the channel in that case.

File: routing/router.go (hunk 4 above)
Comment (joostjager) [reply routing/router.go]: Hmm I see. That is a whole new chapter in failure message interpretation then. I am wondering if it would be better in the end to gradually move those over to local errors that describe the exact problem, instead of a text string `ExtraMsg`.

File: routing/router.go (hunk 10 above)
Comment (joostjager) [reply routing/router.go]: Could be instantiated right above `GetPaymentResult` where it is used. To prevent unnecessary execution if it doesnt get there

Comment (halseth) [reply routing/router.go]: gotcha, fixed here: https://github.com/lightningnetwork/lnd/pull/3063/commits/0a2ff1cb8da7a0e89ca24c8e2efa94e9fa42552f

File: htlcswitch/switch.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (halseth) [reply htlcswitch/switch.go]: making the method return a result async makes it more natural that this is a long-running process. The deadlock happens if the router is waiting for the method to return on shutdown, while the switch will be shutdown after the router.

File: htlcswitch/switch.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (halseth) [reply htlcswitch/switch.go]: Yeah, it is only supported in the upcoming PRs

File: htlcswitch/switch.go (hunk 5 above)
Comment (halseth) [reply htlcswitch/switch.go]: Made a thing: https://github.com/lightningnetwork/lnd/pull/3063/commits/1fa25c266a357dae5f6d758a968df44b5d2eacce

File: htlcswitch/payment_result.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (halseth) [reply htlcswitch/payment_result.go]: I think it would be safe to remove the `isResolution` bool, and just leave the `htlc.Reason==nil` check (or safer `len(htlc.Reason)==0` to handle serialization`)? If there are other cases where the reason can be nil, that is not resolution, decryption would already be failing...

File: htlcswitch/switch.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (halseth) [reply htlcswitch/switch.go]: It was sort of pre-existing, but it wouldn't deadlock on shutdown. Since `router.SendPayment` wouldn't return when shutting down, but the goroutines calling it aren't waitgrouped.

Passing in a quit channel is possible, but not the nicest API imo, as we are not actually cancelling the HTLC or anything, only triggering the call to return.

Having the decryption being done in the router was what the previous iteration of the PR did, but the diff became much larger (especially because the tests and mocks must be changed to return an encrypted error), and we have to pass the error metadata to the router, so we decided to go this route. See: https://github.com/lightningnetwork/lnd/commit/d9ccc3431f47c120d71049ea04f462e4f6d9b7e2

In short terms, it is mostly done to keep the boundary close to the current one, where the `ForwardingError` is returned directly to the router. I do agree it is not optional, but it can be changed later.

File: htlcswitch/switch.go [hunk 19]
```diff
@@ -353,35 +340,90 @@ func (s *Switch) ProcessContractResolution(msg contractcourt.ResolutionMsg) erro
...
+		// Extract the result and pass it to the result channel.
+		result, err := s.extractResult(
+			deobfuscator, n, paymentID, payment.paymentHash,
+		)
+		if err != nil {
+			e := fmt.Errorf("Unable to extract result: %v", err)
+			log.Error(e)
+			resultChan <- &PaymentResult{
+				Error: e,
+			}
```
[the remark this answers is not in the document]
Comment (halseth) [reply htlcswitch/switch.go]: We know the channel is buffered here.

File: htlcswitch/switch.go (hunk 8 above)
Comment (halseth) [reply htlcswitch/switch.go]: Yes, we would need to pass it in, but not sure if it is valuable, for instance in an post-AMP world? (where the switch would only know about payment shards, not the "payment hash").

File: htlcswitch/switch.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (joostjager) [reply htlcswitch/switch.go]: Ok, my suggestion was not to move decryption into the router, but to just to make it a separate call from router to switch. So `switch.SendHTLC` -> `switch.GetPaymentResult` -> `switch.Decrypt` (so slightly different from previous iteration which I remember). But can understand that in this variation the diff in tests may still be large.

File: htlcswitch/switch.go (hunk 8 above)
Comment (joostjager) [reply htlcswitch/switch.go]: I don't know. But if I today see a log statement with a pid in it, does it tell me anything more than no extra info at all? Externally I only know the hash.

File: htlcswitch/switch.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (joostjager) [reply htlcswitch/switch.go]: Ok

File: htlcswitch/switch.go (hunk 8 above)
Comment (halseth) [reply htlcswitch/switch.go]: Added hash back to the log.

File: htlcswitch/payment_result.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (halseth) [reply htlcswitch/payment_result.go]: Decided to keep this as is for now, and decide in the persistence PR.

[the remark this answers is not in the document]
Comment (joostjager) [reply htlcswitch/payment_result.go]: :+1:

File: htlcswitch/switch.go (hunk 11 above)
Comment (cfromknecht) [reply htlcswitch/switch.go]: @halseth nice, i think that is okay. if for some reason this happens twice then it block everything since the channel is only buffered with 1 item. could also just temporarily buffer them at 100 to be safe, but i think the deduping by the circuit map should prevent this.

File: htlcswitch/switch.go [hunk 20]
```diff
@@ -353,35 +340,90 @@ func (s *Switch) ProcessContractResolution(msg contractcourt.ResolutionMsg) erro
...
+		defer s.wg.Done()
+
+		var n *networkResult
+		select {
+		case n = <-payment.resultChan:
+		case <-s.quit:
+			// We close the result channel to signal a shutdown. We
+			// don't send any result in this case since the HTLC is
+			// still in flight.
+			close(resultChan)
```
Comment (cfromknecht) [inline htlcswitch/switch.go]: coudl also make this return `ErrSwitchExiting`, since that is indeed the failure

**Comment by cfromknecht** [comment]
needs rebase as well :)

File: htlcswitch/switch.go (hunk 20 above)
Comment (halseth) [reply htlcswitch/switch.go]: Wanted to keep the result channel reserved for the cases where the HTLC is no longer in flight. Felt it was best to make that distinction clear.

**Comment by halseth** [comment]
Rebased :) PTAL @joostjager @cfromknecht
